### PR TITLE
arch/nrf53: add ADC support

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary
Ported from NRF52

- arch/nrf53: add SAADC support

- boards/nrf5340-dk: add ADC configuration

## Impact
New functionality
## Testing

nrf5340-dk/adc example on the app core

### CODE REVIEW

**Review by acassis — APPROVED** [review]

**Comment by raiden00pl** [comment]
merge conflicts resolved

**Comment by acassis** [comment]
There are new conflicts

**Comment by raiden00pl** [comment]
Conflicts resolved, part two ;]

File: arch/arm/src/nrf53/Kconfig [hunk 1]
```diff
@@ -188,6 +199,64 @@ endchoice
...
+	range 0 8
+	---help---
+		SAADC oversample control
+
+config NRF53_SAADC_RESOLUTION
+	int "SAADC resolution"
+	default 0
+	range 0 3
+	---help---
+		SAADC resolution 0 - 8 bits, 1 - 10 bits, 2 - 12 bits, 3 - 14 bits
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/Kconfig:241]: I think the menu choice would be more user friendly, but we can go with help explanation as initial approach

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 2]
```diff
@@ -0,0 +1,964 @@
...
+ * Name: nrf53_adc_isr
+ *
+ * Description:
+ *   Common ADC interrupt service routine
+ *
+ ****************************************************************************/
+
+static int nrf53_adc_isr(int irq, void *context, void *arg)
+{
+  struct adc_dev_s   *dev  = (struct adc_dev_s *) arg;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:186]:
```suggestion
  struct adc_dev_s   *dev  = (struct adc_dev_s *)arg;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 3]
```diff
@@ -0,0 +1,964 @@
...
+ * Description:
+ *   Bind the upper-half driver callbacks to the lower-half implementation.
+ *   This must be called early in order to receive ADC event notifications.
+ *
+ ****************************************************************************/
+
+static int nrf53_adc_bind(struct adc_dev_s *dev,
+                          const struct adc_callback_s *callback)
+{
+  struct nrf53_adc_s *priv = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:688]:
```suggestion
  struct nrf53_adc_s *priv = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 4]
```diff
@@ -0,0 +1,964 @@
...
+ *
+ * Description:
+ *   Reset the ADC device.  Called early to initialize the hardware.
+ *   This is called, before adc_setup() and on error conditions.
+ *
+ ****************************************************************************/
+
+static void nrf53_adc_reset(struct adc_dev_s *dev)
+{
+  struct nrf53_adc_s *priv = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:709]:
```suggestion
  struct nrf53_adc_s *priv = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 5]
```diff
@@ -0,0 +1,964 @@
...
+ *   Configure the ADC. This method is called the first time that the ADC
+ *   device is opened.  This will occur when the port is first opened.
+ *   This setup includes configuring and attaching ADC interrupts.
+ *   Interrupts are all disabled upon return.
+ *
+ ****************************************************************************/
+
+static int nrf53_adc_setup(struct adc_dev_s *dev)
+{
+  struct nrf53_adc_s *priv = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:732]:
```suggestion
  struct nrf53_adc_s *priv = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 6]
```diff
@@ -0,0 +1,964 @@
...
+ *
+ * Description:
+ *   Disable the ADC.  This method is called when the ADC device is closed.
+ *   This method reverses the operation the setup method.
+ *
+ ****************************************************************************/
+
+static void nrf53_adc_shutdown(struct adc_dev_s *dev)
+{
+  struct nrf53_adc_s *priv = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:805]:
```suggestion
  struct nrf53_adc_s *priv = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 7]
```diff
@@ -0,0 +1,964 @@
...
+ * Name: nrf53_adc_rxint
+ *
+ * Description:
+ *   Call to enable or disable RX interrupts.
+ *
+ ****************************************************************************/
+
+static void nrf53_adc_rxint(struct adc_dev_s *dev, bool enable)
+{
+  struct nrf53_adc_s *priv   = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:833]:
```suggestion
  struct nrf53_adc_s *priv   = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 8]
```diff
@@ -0,0 +1,964 @@
...
+ *
+ * Description:
+ *   All ioctl calls will be routed through this method.
+ *
+ ****************************************************************************/
+
+static int nrf53_adc_ioctl(struct adc_dev_s *dev, int cmd,
+                           unsigned long arg)
+{
+  struct nrf53_adc_s *priv = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:864]:
```suggestion
  struct nrf53_adc_s *priv = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 9]
```diff
@@ -0,0 +1,964 @@
...
+    }
+#endif
+
+  /* Get device */
+
+  dev = &g_nrf53_adc;
+
+  /* Get private data */
+
+  priv = (struct nrf53_adc_s *) dev->ad_priv;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:946]:
```suggestion
  priv = (struct nrf53_adc_s *)dev->ad_priv;
```

File: arch/arm/src/nrf53/nrf53_adc.c [hunk 10]
```diff
@@ -0,0 +1,964 @@
...
+  int                 i    = 0;
+
+  DEBUGASSERT(chan != NULL);
+  DEBUGASSERT(channels <= CONFIG_NRF53_SAADC_CHANNELS);
+
+#ifdef CONFIG_NRF53_SAADC_TIMER
+  if (channels > 1)
+    {
+      aerr("ERORR: timer trigger works only for 1 channel!\n");
+      goto errout;
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.c:936]: Why not just `return NULL;`?

File: arch/arm/src/nrf53/nrf53_adc.h [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+  uint16_t limitl;              /* Low limit */
+#endif
+  uint8_t resp:2;               /* Positive chan resistor */
+  uint8_t resn:2;               /* Negative chan resistor */
+  uint8_t gain:3;               /* Gain control */
+  uint8_t refsel:1;             /* Reference control */
+  uint8_t tacq:3;               /* Acquisition time */
+  uint8_t mode:1;               /* Singe-ended or differential mode */
+  uint8_t burst:1;              /* Burst mode */
+  uint8_t _res:3;               /* Reserved */
```
Comment (pkarashchenko) [inline arch/arm/src/nrf53/nrf53_adc.h:136]:
```suggestion
  uint16_t resp:2;              /* Positive chan resistor */
  uint16_t resn:2;              /* Negative chan resistor */
  uint16_t gain:3;              /* Gain control */
  uint16_t refsel:1;            /* Reference control */
  uint16_t tacq:3;              /* Acquisition time */
  uint16_t mode:1;              /* Singe-ended or differential mode */
  uint16_t burst:1;             /* Burst mode */
  uint16_t _res:3;              /* Reserved */
```

File: boards/arm/nrf53/nrf5340-dk/src/nrf53_adc.c [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+  /* Channel 3 */
+
+  {
+    .p_psel  = NRF53_ADC_IN_IN5,
+    .n_psel  = 0,
+#ifdef CONFIG_NRF53_SAADC_LIMITS
+    .limith = 0,
+    .limitl = 0,
+#endif
```
Comment (pkarashchenko) [inline boards/arm/nrf53/nrf5340-dk/src/nrf53_adc.c:132]: this pars is not actually needed

File: boards/arm/nrf53/nrf5340-dk/src/nrf53_adc.c [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+      for (i = 0; i < ADC_NCHANNELS; i += 1)
+        {
+          nrf53_gpio_config(g_adc_pins[i]);
+        }
+
+      /* Call nrf53_adcinitialize() to get an instance of the ADC interface */
+
+      adc = nrf53_adcinitialize(g_adc_chanlist, ADC_NCHANNELS);
+      if (!adc)
```
Comment (pkarashchenko) [inline boards/arm/nrf53/nrf5340-dk/src/nrf53_adc.c:177]: Optional
```suggestion
      if (adc == NULL)
```

**Comment by raiden00pl** [comment]
@pkarashchenko at the moment I would prefer to leave everything as it is for compliance with the NRF52. Later I can update both chips at once

**Comment by raiden00pl** [comment]
conflicts resolved

**Comment by pkarashchenko** [comment]
Sure. All is fine like it is now. My comments were optional

File: arch/arm/src/nrf53/nrf53_adc.c (hunk 10 above)
Comment (pkarashchenko) [reply arch/arm/src/nrf53/nrf53_adc.c:936]: If we return `NULL` here, then we can eliminate `errout` label

**Pull request merged**